File: 01-Programming-Languages/TypeScript/Data-Structures/LinkedList/linkedListCycle.ts
/*
Problem: Linked List Cycle
Description: Detect if a linked list has a cycle using Floyd's algorithm.

Approach:
- Use slow and fast pointers; if they meet, a cycle exists

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input: 3 -> 2 -> 0 -> -4 (cycle at node with value 2)
Output: true
*/

class ListNode {
  constructor(public val: number, public next: ListNode | null = null) {}
}

function hasCycle(head: ListNode | null): boolean {
  let slow: ListNode | null = head, fast: ListNode | null = head;
  while (fast && fast.next) {
    slow = slow!.next;
    fast = fast.next.next;
    if (slow === fast) return true;
  }
  return false;
}

const node1 = new ListNode(3);
const node2 = new ListNode(2);
const node3 = new ListNode(0);
const node4 = new ListNode(-4);
node1.next = node2;
node2.next = node3;
node3.next = node4;
node4.next = node2;

const noCycle = new ListNode(1, new ListNode(2));

console.log('Has cycle (with cycle):', hasCycle(node1));
console.log('Has cycle (no cycle):', hasCycle(noCycle));

File: 01-Programming-Languages/C++/Data-Structures/LinkedList/linked_list_cycle.cpp
/*
Problem: Linked List Cycle
Detect if a linked list has a cycle using Floyd's cycle detection.

Approach:
- Two pointers (slow and fast)

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input: 1 -> 2 -> 3 -> 4 -> 2 (cycle)
Output: Cycle detected
*/

#include <iostream>
using namespace std;

struct Node
{
    int data;
    Node *next;
    Node(int val) : data(val), next(nullptr) {}
};

bool has_cycle(Node *head)
{
    Node *slow = head, *fast = head;
    while (fast && fast->next)
    {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) return true;
    }
    return false;
}

int main()
{
    Node *head = new Node(1);
    head->next = new Node(2);
    head->next->next = new Node(3);
    head->next->next->next = new Node(4);
    head->next->next->next->next = head->next; // create cycle
    cout << (has_cycle(head) ? "Cycle detected" : "No cycle") << endl;
    return 0;
}
